### PR TITLE
Use skaffold to run agent in k8s

### PR DESCRIPTION
**Pull request by pchila**
## What does this PR do?

Creates a skaffold config along with the necessary bits for (quickly) creating a docker image and manifest transformation for running/debugging agent on a local k8s cluster.

Two profiles `elastic-agent-standalone` and `elastic-agent-managed` have been created to run the agent in both modes.

*Note*: this has been tested with [kind](https://kind.sigs.k8s.io/) and [k3d](https://k3d.io/) but it should work with other local k8s cluster. It can be used also with external clusters, although the details about pushing the agent image on remote docker registries have not yet been worked out.

*Note 2*: the build of the elastic agent image is very hacky at the moment, a better way should be implemented before this change is merged

## Checklist

- ~~[ ] My code follows the style guidelines of this project~~
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- ~~[ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- ~~[ ] I have added an integration test or an E2E test~~

## Author's Checklist

- [ ]

## How to test this PR locally

- Install [skaffold](https://skaffold.dev/) and a local k8s cluster of your choice (kind is recommended)
- Fill the necessary env vars in `deploy/skaffold/.env`, using the example below:
  ```
  FLEET_URL=https://<my fleet host>:443
  FLEET_ENROLLMENT_TOKEN=<my enrollment tokebn>
  ES_HOST=https://<my elasticsearch host>:443
  ES_USERNAME=<elasticsearch username>
  ES_PASSWORD=<elasticsearch password>
  ```
- Make sure that your k8s context is pointing to the right (local) cluster
- Run the agent by running `skaffold run -p [elastic-agent-standalone|elastic-agent-managed]` or by using the [google cloud code extension](https://cloud.google.com/code) for your IDE
- Wait for the image to be built and deployed on your cluster
- If you want to debug your code you can run `skaffold debug -p [elastic-agent-standalone|elastic-agent-managed]`. This command modifies the elastic-agent container by adding `dlv` debugger and starting the elastic-agent binary with it so a `dlv` client can be connected to it via port forwarding (provided by default)

## Related issues

- 

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

**Comment by michalpristas** [comment]
@pchila are you going to address what we discussed during sync or is this considered final?

**Review by belimawr — COMMENTED** [review]
The only thing I'm missing is some entry in the developer documentation about how to use this PR, especially mentioning the `.env` files, they are hidden on Unix-like systems, so it will be hart to know they need to be edited.

Even a simple `Readme.md` on `deploy/skaffold` will do.

Nit: many files are missing the blank line at the end.

**Review by belimawr — CHANGES_REQUESTED** [review]
Updating my review to 'request changes'. We really need some small docs about how to use this.

File: docs/local-k8s-testing.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+### Prepare environment variables
+In order to run agent on a local k8s cluster we need to set some environment variables in `deploy/skaffold/.env` (we can use `.env.example`). Those environment variables must point to a running elastic stack installation.
+
+#### Standalone mode
+```shell
+# standalone elastic agent vars
+ES_HOST=https://<elasticsearch host>:443
+ES_USERNAME=elastic
+ES_PASSWORD=changeme
+
```
Comment (belimawr) [inline docs/local-k8s-testing.md:30]:
```suggestion
```

**Review by belimawr — APPROVED** [review]

**Comment by pchila** [comment]
@michalpristas  the upload of images to a remote image registry works however docker.elastic.co is not a suitable registry as it's not possible to assign a ttl or a cleanup policy to the images we upload there and deleting images is not sokmething that can be done self-service. For the moment we are going to merge this pr as-is just with local k8s cluster support

File: Dockerfile.skaffold [hunk 2]
```diff
@@ -0,0 +1,27 @@
+ARG GO_VERSION=1.19.9
+ARG crossbuild_image="docker.elastic.co/beats-dev/golang-crossbuild"
+ARG AGENT_VERSION=8.9.0-SNAPSHOT
```
Comment (michalpristas) [inline Dockerfile.skaffold:3]: we need to update this with each version bump. 
do you think we can read a file and inject it here?

File: Dockerfile.skaffold [hunk 3]
```diff
@@ -0,0 +1,27 @@
+ARG GO_VERSION=1.19.9
```
Comment (michalpristas) [inline Dockerfile.skaffold:1]: same as agent version, would be good to have this loaded from .go-version

**Review by michalpristas — APPROVED** [review]
lgtm, please think about options to have go and beat version injected magically

**Comment by pchila** [comment]
@michalpristas  I will have a look at the options for updating the go version and the base elastic agent image version in a follow-up PR. Opened https://github.com/elastic/elastic-agent/issues/2960 to track this

**Pull request merged**